src/Compilers/CSharp/Test/Emit/SourceGeneratorTests.cs

```diff
@@ -0,0 +1,201 @@
+        {
+            string text =
+@"class C
+{
+}";
+            using (var directory = new DisposableDirectory(Temp))
+            {
+                var file = directory.CreateFile("c.cs");
+                file.WriteAllText(text);
+                var diagnostics = DiagnosticBag.GetInstance();
+                Assert.Throws<OperationCanceledException>(() =>
+                    RunCompiler(
+                        directory.Path,
+                        file.Path,
+                        diagnostics,
+                        ImmutableArray.Create<SourceGenerator>(
+                            new SimpleSourceGenerator(c => { throw new InvalidCastException(); }),
+                            new SimpleSourceGenerator(c => { throw new OperationCanceledException(); }))));
+                diagnostics.Verify();
+                diagnostics.Free();
+            }
+        }
+
+        private static void RunCompiler(string baseDirectory, string filePath, DiagnosticBag diagnostics, ImmutableArray<SourceGenerator> generators)
+        {
+            var compiler = new MyCompiler(
+                baseDirectory: baseDirectory,
+                args: new[] { "/nologo", "/preferreduilang:en", "/t:library", filePath },
+                generators: generators);
+            var errorLogger = new DiagnosticBagErrorLogger(diagnostics);
+            var builder = new StringBuilder();
+            using (var writer = new StringWriter(builder))
+            {
+                compiler.RunCore(writer, errorLogger, default(CancellationToken));
+            }
+        }
+
+        private sealed class MyCompiler : CSharpCompiler
+        {
+            private readonly ImmutableArray<SourceGenerator> _generators;
+
+            internal MyCompiler(
+                string baseDirectory,
+                string[] args,
+                ImmutableArray<SourceGenerator> generators) :
+                base(
+                    new CSharpCommandLineParser(),
+                    responseFile: null,
+                    args: args,
+                    clientDirectory: null,
+                    baseDirectory: baseDirectory,
+                    sdkDirectoryOpt: RuntimeEnvironment.GetRuntimeDirectory(),
+                    additionalReferenceDirectories: null,
+                    assemblyLoader: null)
+            {
+                _generators = generators;
+            }
+
+            protected override void ResolveAnalyzersAndGeneratorsFromArguments(
+                List<DiagnosticInfo> diagnostics,
+                CommonMessageProvider messageProvider,
+                out ImmutableArray<DiagnosticAnalyzer> analyzers,
+                out ImmutableArray<SourceGenerator> generators)
+            {
+                analyzers = ImmutableArray<DiagnosticAnalyzer>.Empty;
+                generators = _generators;
+            }
+
+            protected override void CompilerSpecificSqm(IVsSqmMulti sqm, uint sqmSession)
```

Comment:
I think the plan is to rip out SQM. You can leave the stubs if you want, or just delete the definitions.